django_admin_lightweight_date_hierarchy/admin.py

```diff
@@ -0,0 +1,86 @@
+import re
+import datetime
+
+from django.core.exceptions import ImproperlyConfigured
+from django.conf import settings
+from django.contrib import admin
+from django.utils import timezone
+
+
+def get_date_range_for_hierarchy(date_hierarchy, tz):
+    """Generate date range for date hierarchy.
+
+    date_hierarchy <dict>:
+        year (int or None)
+        month (int)
+        day (int)
+    tz <timezone or None>:
+        The timezone in which to generate the datetimes.
+        If None, the datetimes will be naive.
+
+    Returns (tuple):
+        from_date (datetime.datetime, aware if tz is set) inclusive
+        to_date (datetime.datetime, aware if tz is set) exclusive
+    """
+    from_date = datetime.datetime(
+        date_hierarchy['year'],
+        date_hierarchy.get('month', 1),
+        date_hierarchy.get('day', 1),
+    )
+
+    if tz:
+        from_date = tz.localize(from_date)
+
+    if 'day' in date_hierarchy:
+        to_date = from_date + datetime.timedelta(days=1)
+
+    elif 'month' in date_hierarchy:
+        assert from_date.day == 1
+        to_date = (from_date + datetime.timedelta(days=32)).replace(day=1)
+
+    else:
+        to_date = from_date.replace(year=from_date.year + 1)
+
+    return from_date, to_date
+
+
+class EnhancedDateHierarchyListFilter(admin.ListFilter):
```

Comment:
"Enhanced" is pretty generic. I think a name like `RangeBasedDateHierarchyListFilter` would be better.